packages/core/src/submodules/cbor/cbor-decode.ts

```diff
@@ -0,0 +1,477 @@
+            const countOffset = (countLength + 1) as CborArgumentLengthOffset;
+            offset = countOffset;
+
+            if (to - at < countOffset) {
+              throw new Error(`countLength ${countLength} greater than remaining buf len.`);
+            }
+            const countIndex = at + 1;
+            if (countLength === 1) {
+              unsignedInt = payload[countIndex];
+            } else if (countLength === 2) {
+              unsignedInt = dataView.getUint16(countIndex);
+            } else if (countLength === 4) {
+              unsignedInt = dataView.getUint32(countIndex);
+            } else {
+              unsignedInt = dataView.getBigUint64(countIndex);
+            }
+            break;
+          default:
+            throw new Error(`unexpected minor value ${minor}.`);
+        }
+      }
+
+      if (major === majorUint64) {
+        _offset = offset;
+        return castBigInt(unsignedInt);
+      } else if (major === majorNegativeInt64) {
+        let negativeInt: bigint | number;
+        if (typeof unsignedInt === "bigint") {
+          negativeInt = BigInt(-1) - unsignedInt;
+        } else {
+          negativeInt = -1 - unsignedInt;
+        }
+        _offset = offset;
+        return castBigInt(negativeInt);
+      } else {
+        const value = decode(at + offset, to);
+        const valueOffset = _offset;
+
+        _offset = offset + valueOffset;
+        return { tag: castBigInt(unsignedInt), value };
+      }
+    case majorUtf8String:
+    case majorMap:
+    case majorList:
+    case majorUnstructuredByteString:
+      if (minor === minorIndefinite) {
+        switch (major) {
+          case majorUtf8String:
+            return decodeUtf8StringIndefinite(at, to);
+          case majorMap:
+            return decodeMapIndefinite(at, to);
+          case majorList:
+            return decodeListIndefinite(at, to);
+          case majorUnstructuredByteString:
+            return decodeUnstructuredByteStringIndefinite(at, to);
+        }
+      } else {
+        switch (major) {
+          case majorUtf8String:
+            return decodeUtf8String(at, to);
+          case majorMap:
+            return decodeMap(at, to);
+          case majorList:
+            return decodeList(at, to);
+          case majorUnstructuredByteString:
+            return decodeUnstructuredByteString(at, to);
+        }
+      }
+    default:
+      return decodeSpecial(at, to);
+  }
+}
+
+function bytesToUtf8(bytes: Uint8Array, at: number, to: number): string {
+  if (USE_BUFFER && bytes.constructor?.name === "Buffer") {
+    return (bytes as Buffer).toString("utf-8", at, to);
+  }
+  if (textDecoder) {
+    return textDecoder!.decode(bytes.subarray(at, to));
+  }
+  return toUtf8(bytes.subarray(at, to));
+}
+
+function demote(bigInteger: bigint): number {
+  const num = Number(bigInteger);
+  if (num < Number.MIN_SAFE_INTEGER || Number.MAX_SAFE_INTEGER < num) {
```

Comment:
This function is used to demote (cast) a `BigUint64` to `number`. 
This is not _generally_ safe, but in this context it is safe because it is only used to retrieve a string or array length.

Since `Number.MAX_SAFE_INTEGER`  is much larger than the spec-allowed maximums for JS strings and arrays, this will not cause information loss.